src/gs-update-monitor.c

```diff
@@ -150,6 +190,269 @@ no_updates_for_a_week (GsUpdateMonitor *monitor)
+
+	if (!ret) {
+		g_warning ("Failed scheduled update of %s: %s",  app_id,
+		           local_error->message);
+		return;
+	}
+
+	g_debug ("Scheduled update of app %s succeeded", app_id);
+
+	/* unschedule the update */
+	g_hash_table_remove (helper->monitor->scheduled_updates, app_id);
+}
+
+static void
+update_app (UpdateScheduleHelper *helper)
+{
+	GsApp *app = helper->app;
+	g_autoptr(GsPluginJob) plugin_job = NULL;
+
+	g_debug ("Performing scheduled update for app %s",
+		 gs_app_get_unique_id (app));
+
+	app_set_auto_updating (app, TRUE);
+
+	plugin_job = gs_plugin_job_newv (GS_PLUGIN_ACTION_UPDATE,
+					 "app", app,
+					 "failure-flags", GS_PLUGIN_FAILURE_FLAGS_NONE,
+					 NULL);
+	gs_plugin_loader_job_process_async (helper->monitor->plugin_loader,
+					    plugin_job, NULL,
+					    (GAsyncReadyCallback) app_update_finished_cb,
+					    helper);
+}
+
+static void
+download_now_cb (GObject *obj,
+                 GParamSpec *pspec,
+                 UpdateScheduleHelper *helper)
+{
+	GsApp *app = helper->app;
+	gboolean download_now = mwsc_schedule_entry_get_download_now (helper->entry);
+	AsAppState state = gs_app_get_state (helper->app);
+	const gchar *app_id = gs_app_get_unique_id (app);
+
+	g_debug ("Got download-now=%s for updating app %s",
+		 (download_now ? "TRUE" : "FALSE"), app_id);
+
+	if (download_now) {
+		/* verify again if the app needs to be updated, if not,
+		 * unschedule the update */
+		if (state == AS_APP_STATE_UPDATABLE_LIVE) {
+			update_app (helper);
+		} else {
+			g_debug ("Should update app %s but its state is %s! "
+				 "Unscheduling the update...", app_id,
+				 as_app_state_to_string (state));
+
+			/* unschedule the update */
+			g_hash_table_remove (helper->monitor->scheduled_updates,
+					     app_id);
+		}
+		return;
+	}
+
+	if (state == AS_APP_STATE_INSTALLING) {
+		/* if we cannot update at the moment, cancel any automatically
+		 * started update */
+		if (!app_is_auto_updating (helper->app))
+			return;
+
+		g_debug ("Cancelling automatic update of app %s, as "
+			 "download-now is FALSE", app_id);
+		g_cancellable_cancel (gs_app_get_cancellable (helper->app));
+	}
+}
+
+static void
+monitor_clear_scheduler (GsUpdateMonitor *monitor)
+{
+	g_clear_object (&monitor->scheduler);
+}
+
+static void
+async_result_cb (GObject *source_object,
+                 GAsyncResult *result,
+                 GAsyncResult **out_result)
+{
+	*out_result = g_object_ref (result);
+}
+
+static gboolean
+setup_scheduler (GsUpdateMonitor *monitor,
+		 GCancellable *cancellable,
+		 GError **error)
+{
+	g_autoptr(GAsyncResult) new_result = NULL;
+	GMainContext *context = NULL;
+	MwscScheduler *scheduler = NULL;
+
+	if (monitor->scheduler != NULL)
+		return TRUE;
+
+	context = g_main_context_default ();
+
+	mwsc_scheduler_new_async (cancellable,
+				  (GAsyncReadyCallback) async_result_cb,
+				  &new_result);
+	while (new_result == NULL)
+		g_main_context_iteration (context, TRUE);
```

Comment:
Same comment to the one I made on `schedule_update` here: I think this is wrong since you're manually spinning the main loop of the global main context manually. Either properly deal with asynchrony here or move it into its own thread (and probably use `g_main_context_get_thread_default`)

Reply:
Why is this a problem? The `GsUpdateMonitor` runs in the main thread. The first version of this code was using the real asynchronicity as you say, and I usually prefer that (iterating in loops is ugly), but the code did get more complicated, so added it like this.

Reply:
The problem is that you're iterating the main loop of the global context, which might be processing other sources than the ones you're interested in for the purpose of this functionality. If you want to go the "manually spinning the loop" route, I think a better way would be to create a new context and push it as the thread default before starting the asynchronous operation, and then spinning it from there, then pop that context out.

Reply:
What Mario said. Additionally, how can you be sure the `GsUpdateMonitor` will always run in the main thread? If the code gets refactored in the future, this will silently start to cause a lot of hard to debug problems.